text/0000-unsized-rvalues.md

```diff
@@ -0,0 +1,203 @@
+- Feature Name: unsized_locals
+- Start Date: 2017-02-11
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow for local variables, function arguments, and some expressions to have an unsized type, and implement it by storing the temporaries in variably-sized allocas.
+
+Have repeat expressions with a length that captures local variables be such an expression, returning an `[T]` slice.
+
+Provide some optimization guarantees that unnecessary temporaries will not create unnecessary allocas.
+
+# Motivation
+[motivation]: #motivation
+
+There are 2 motivations for this RFC:
+
+1. Passing unsized values, such as trait objects, to functions by value is often desired. Currently, this must be done through a `Box<T>` with an unnecessary allocation.
+
+  One particularly common example is passing closures that consume their environment without using monomorphization. One would like for this code to work:
+
+  ```Rust
+  fn takes_closure(f: FnOnce()) { f(); }
+  ```
+
+  But today you have to use a hack, such as taking a `Box<FnBox<()>>`.
+
+2. Allocating a runtime-sized variable on the stack is important for good performance in some use-cases - see RFC #1808, which this is intended to supersede.
+
+# Detailed design
+[design]: #detailed-design
+
+## Unsized Rvalues - language
+
+Remove the rule that requires all locals and rvalues to have a sized type. Instead, require the following:
+
+1. The following expressions must always return a Sized type:
+    1. Function calls, method calls, operator expressions
+        - implementing unsized return values for function calls would require the *called function* to do the alloca in our stack frame.
+    2. ADT expressions
+        - see alternatives
+    3. cast expressions
+        - this seems like an implementation simplicity thing. These can only be trivial casts.
+2. The RHS of assignment expressions must always have a Sized type.
+    - Assigning an unsized type is impossible because we don't know how much memory is available at the destination. This applies to ExprAssign assignments and not to StmtLet let-statements.
+
+This also allows passing unsized values to functions, with the ABI being as if a `&move` pointer was passed (a `(by-move-data, extra)` pair). This also means that methods taking `self` by value are object-safe, though vtable shims are sometimes needed to translate the ABI (as the callee-side intentionally does not pass `extra` to the fn in the vtable, no vtable shim is needed if the vtable function already takes its argument indirectly).
```

Comment:
Suppose you have something like:
```Rust
trait Foo {
    fn bar(self, x: u32);
}

struct Bar(...);

impl Foo for Bar { /* .. */ }
```

The function (shim) `<Bar as Foo>::foo` takes 3 "LLVM" argument - a fat by-move pointer to `Self`, and `x` by-value.

```
function <Bar as Foo>::foo(%data: *u8, %vtbl: *<vtable for Foo>, %x: u32) {
    %fn = vtbl.foo;
    call %fn(%data, %x); // vtbl is not passed.
}
```

Then the vtable entry for `<Bar as Foo>::foo` has the type `llfn(*u8, u32)`. The Rust function `<Bar as Foo>::foo` has the Rust type `fn(Foo, u32)`. If `Foo` is a large struct that is passed by reference, this is lowered to the llvm type `llfn(*Foo, u32)` which can be bitcast. OTOH, if `Foo` is, say, an u32, this is lowered to `llfn(u32, u32)` and we need a shim that passes a reference.